hack/docgen/schema.go

```diff
@@ -0,0 +1,242 @@
+		}
+	}
+
+	if strings.HasPrefix(goType, "[]") {
+		return SchemaDefinitionInfo{
+			typeInfo:           SchemaTypeInfo{typeName: "array"},
+			arrayItemsTypeInfo: goTypeToTypeInfo(strings.TrimPrefix(goType, "[]")),
+		}
+	}
+
+	if strings.HasPrefix(goType, "map[string]") {
+		return SchemaDefinitionInfo{
+			typeInfo:         SchemaTypeInfo{typeName: "object"},
+			mapValueTypeInfo: goTypeToTypeInfo(strings.TrimPrefix(goType, "map[string]")),
+		}
+	}
+
+	return SchemaDefinitionInfo{
+		typeInfo: *goTypeToTypeInfo(goType),
+	}
+}
+
+func typeInfoToSchema(typeInfo *SchemaTypeInfo) *jsonschema.Schema {
+	schema := &jsonschema.Schema{}
+
+	if typeInfo.typeName != "" {
+		schema.Type = typeInfo.typeName
+	}
+
+	if typeInfo.ref != "" {
+		schema.Ref = typeInfo.ref
+	}
+
+	return schema
+}
+
+func fieldToSchema(field *Field) *jsonschema.Schema {
+	schema := jsonschema.Schema{}
+
+	if field.Text != nil {
+		// if there is an explicit schema, use it
+		if field.Text.Schema != nil {
+			schema = field.Text.Schema.Schema
+		}
+
+		// if no title is provided on the explicit schema, grab it from the comment
+		if schema.Title == "" {
+			schema.Title = strings.ReplaceAll(field.Tag, "\\n", "\n")
+		}
+
+		// if no description is provided on the explicit schema, grab it from the comment
+		if schema.Description == "" {
+			schema.Description = normalizeDescription(field.Text.Description)
+		}
+
+		// if an explicit schema was provided, return it
+		if field.Text.Schema != nil {
+			return &schema
+		}
+	}
+
+	// schema was not explicitly provided, generate it from the comment
+
+	info := fieldToDefinitionInfo(field)
+
+	if info.typeInfo.ref != "" {
+		schema.Ref = info.typeInfo.ref
+	}
+
+	if info.enumValues != nil {
+		schema.Enum = info.enumValues
+	}
+
+	if info.typeInfo.typeName != "" {
+		schema.Type = info.typeInfo.typeName
+	}
+
+	if info.arrayItemsTypeInfo != nil {
+		schema.Items = typeInfoToSchema(info.arrayItemsTypeInfo)
+	}
+
+	if info.mapValueTypeInfo != nil {
+		schema.PatternProperties = map[string]*jsonschema.Schema{
+			".*": typeInfoToSchema(info.mapValueTypeInfo),
+		}
+	}
+
+	return &schema
+}
+
+func structToSchema(st *Struct) *jsonschema.Schema {
+	schema := &jsonschema.Schema{
+		Type:                 "object",
+		AdditionalProperties: jsonschema.FalseSchema,
+	}
+
+	properties := orderedmap.New()
+
+	for _, field := range st.Fields {
+		if field.Tag == "" {
+			// skip unknown/untagged field
+			continue
+		}
+		properties.Set(field.Tag, fieldToSchema(field))
+	}
+
+	schema.Properties = properties
+
+	return schema
+}
+
+func docToSchema(doc *Doc) *jsonschema.Schema {
+	schema := &jsonschema.Schema{
+		Version: jsonschema.Version,
+		ID:      ConfigSchemaURL,
+		Ref:     "#/$defs/Config",
+	}
+
+	definitions := make(map[string]*jsonschema.Schema, len(doc.Structs))
+
+	for _, st := range doc.Structs {
+		definitions[st.Name] = structToSchema(st)
+	}
+
+	schema.Definitions = definitions
```

Comment:
I think we can use [slices.ToMap](https://github.com/siderolabs/gen/blob/main/slices/slices.go#L89) here directly

Reply:
done